### ISSUE TITLE
trial should support color output on win32 as well

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11153]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#1646|
|Type|enhancement|
|Created|2006-04-16 12:45:24Z|

It's useless fluff, but hey, it should be _portable_ useless fluff.

I will attach some stuff.

Attachments:

* [colortrial.png](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc0/cc0152ff71fc8243f70c8e8478bd7de2aa387e2c/8712ff73168ebacd1743c8441603e29f9d3227c2.png) (5087 bytes) - added by glyph on 2006-04-16 12:46:22Z - 
* [colortrial.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cc0/cc0152ff71fc8243f70c8e8478bd7de2aa387e2c/bc309f9768d260883bef76301b54ec159dc4ecf3.py) (1034 bytes) - added by glyph on 2006-04-16 12:46:33Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1646 1646
type__enhancement enhancement
reporter__glyph glyph
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__win32 win32
time__1145191524000000 1145191524000000
changetime__1154440866000000 1154440866000000
version__ 
owner__ 
cc__teratorn cc__exarkun cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

I don't have a win32 machine, so I don't really know what to do with this.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

I've decided what to do with this. 

Give me access to a win32 development machine.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn commented|
|-|-|

Done!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Ready for review in [branches/win32color-1646](branches/win32color-1646)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @teratorn<br>@jml set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)|@teratorn set owner to @jml|
|-|-|

Review OK - makes sense and works in my test environments.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `closed`|
|-|-|

(In [17733]) Make Trial work with color on win32 platforms.

 * Author: jml
 * Reviewer: teratorn
 * Fixes [#11153](https://github.com/twisted/twisted/issues/11153)

Extracts the color stuff from TreeReporter into separate classes.
This makes it easy to add color support for exotic platforms like 
windows. 

Before this branch, if a terminal had no color support, Trial fell
back to the VerboseTextReporter. Now, it will still use the 
TreeReporter -- it will just print without colour.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

There was not even one docstring or unit test added in this branch :(

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

(In [17758]) Revert r17733 - basic functionality regression

Refs [#11153](https://github.com/twisted/twisted/issues/11153)

Reporting of cleanup errors results in an unhandled exception in trial when using
the colorized reporter with this changeset.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @teratorn<br>@jml set status to `new`|
|-|-|

Fixed the failure, added a test and some docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml<br>@exarkun set status to `new`|
|-|-|

For MockColorizer, wouldn't it be simpler to keep a list of two-tuples instead of a StringIO of formatted data?  Then all the string splitting in test_cleanupError can be removed.

Isn't test_reporter.suppressWarnings(foo) the same as twisted.internet.util.suppressWarnings([twisted.trial.util.suppress(category=foo)])?

In any case, the import should be outside the definition rather than inside.  Likewise for _AnsiColorizer.supported's imports of sys and os.

colorizer.supported and colorizer.write are part of an internal interface, but it's one a future maintainer might want to implement or expand - add docstrings for these, please.

_AnsiColorizer.write interpolates literal strings ('\x1b[').  Not wrong, but weird, since it also includes parts of the ecma control sequence in the format string.

Thanks for adding the test.  Also, the change from ```self.GREEN``` to ```self.SUCCESS``` etc is nice. :)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `closed`|
|-|-|

(In [17781]) Make Trial work with color on win32 platforms.

    * Author: jml
    * Reviewers: exarkun, teratorn
    * Fixes [#11153](https://github.com/twisted/twisted/issues/11153) 

Extracts the color stuff from `TreeReporter` into separate classes. 
This makes it easy to add color support for exotic platforms like 
windows.

Before this branch, if a terminal had no color support, Trial fell 
back to the `VerboseTextReporter`. Now, it will still use the 
`TreeReporter` -- it will just print without colour. 

UPDATE: Fixed the regression by adding a test (yay!). Also added 
docstrings and cleaned things up a little.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
